Rad2deg

leoheim opened the pull request:
### Details:
 - *Added support and test case files for aten::rad2deg *

### Tickets:
 - *thttps://github.com/openvinotoolkit/openvino/issues/28873*

rkazants commented on src/frontends/pytorch/src/op/rad2deg.cpp:
```diff
@@ -0,0 +1,39 @@
+#include "openvino/frontend/pytorch/node_context.hpp"
+#include "openvino/op/multiply.hpp"
+#include "openvino/op/constant.hpp"
+#include <cmath>  // Required for M_PI
+
+#ifndef M_PI
+#define M_PI 3.14159265358979323846
+#endif
```
please remove it. Instead of this, please use this code:

```
const double pi_val = std::atan(1.0) * 4;
```

rkazants commented on src/frontends/pytorch/src/op/rad2deg.cpp:
```diff
@@ -0,0 +1,39 @@
+    auto input = context.get_input(0);
+    
+    // Create a constant node with the conversion factor (180 / π)
+    auto conversion_factor = context.mark_node(ov::op::v0::Constant::create(element::f32, Shape{}, {180.0 / M_PI}));
```
please align type of `conversion_factor`. It should be of type `input`

rkazants commented on src/frontends/pytorch/src/op_table.cpp:
```diff
@@ -340,11 +341,13 @@ OP_CONVERTER(translate_conv1d_ext);
 OP_CONVERTER(translate_embedding_ext);
 OP_CONVERTER(translate_linear_awq);
 
+
```
```suggestion
```

leoheim commented:
Hey @rkazants , I have made the requested changes. Let me know if anything else is needed. Thanks

leoheim commented:
Hi @rkazants ,

I hope you're doing well! Just wanted to follow up on my PR whenever you have a moment to review them. I’d appreciate any feedback or suggestions.

Thank you!

mvafin commented on src/frontends/pytorch/src/op_table.cpp:
```diff
@@ -340,11 +341,13 @@ OP_CONVERTER(translate_conv1d_ext);
 // Supported ops for TorchScript
 const std::unordered_map<std::string, CreatorFunction> get_supported_ops_ts() {
     return {
+        {"aten::rad2deg", op::translate_rad2deg},
```
Please move this to preserve alphabetical order

mvafin commented on src/frontends/pytorch/src/op_table.cpp:
```diff
@@ -19,6 +19,7 @@ namespace op {
 using namespace ov::frontend;
 
 // TorchScript translations
+OP_CONVERTER(translate_rad2deg);
```
Please move this to preserve alphabetical order

mvafin reviewed (changes requested):
Please fix code style

leoheim commented:
Hi @mvafin,

I’ve made the requested changes. Let me know if there’s anything else needed!

leoheim commented:
Hello @mvafin,

I’ve made the requested changes. Let me know if there’s anything else needed!

mvafin commented:
Code style still fails:

```
src/frontends/pytorch/src/op/rad2deg.cpp:8:-#include "openvino/op/multiply.hpp"
src/frontends/pytorch/src/op/rad2deg.cpp:9:+#include "openvino/op/multiply.hpp"
src/frontends/pytorch/src/op/rad2deg.cpp:19:-    
src/frontends/pytorch/src/op/rad2deg.cpp:19:+
src/frontends/pytorch/src/op/rad2deg.cpp:27:-    
src/frontends/pytorch/src/op/rad2deg.cpp:27:+
src/frontends/pytorch/src/op/rad2deg.cpp:30:-    
src/frontends/pytorch/src/op/rad2deg.cpp:30:+
src/frontends/pytorch/src/op_table.cpp:354:-
```

You can run clang-format to check for style issues

mvafin commented on src/frontends/pytorch/src/op/rad2deg.cpp:
```diff
@@ -0,0 +1,41 @@
+
+OutputVector translate_rad2deg(const NodeContext& context) {
+    // Ensure that the operation has exactly one input (the input tensor)
+    num_inputs_check(context, 1, 1);
```
Add #include "utils.hpp" for this function

leoheim commented:
@mvafin Included now. Thanks!

mvafin commented on src/frontends/onnx/frontend/src/op/max_pool.cpp:
why this change?

mvafin commented on src/common/transformations/src/transformations/convert_precision.cpp:
why this change?

leoheim commented:
Hi @mvafin, I apologize for the confusion. It seems I accidentally mixed changes from another branch into this one. I've corrected the issue now, and the branch should be clean. Thank you for pointing this out!

mvafin commented:
@leoheim  Please add test. Seems like you removed it by accident

leoheim commented:
@mvafin I have added the test now. Apologies for the confusion earlier!

mvafin commented on tests/layer_tests/pytorch_tests/test_rad2deg.py:
```diff
@@ -0,0 +1,26 @@
+            def forward(self, x):
+                return x * (180.0 / np.pi)
+        ref_net = None  
+        return aten_rad2deg(), ref_net, "aten::mul"
```
This is not a test for rad2deg
```suggestion
        return aten_rad2deg(), ref_net, "aten::rad2deg"
```

mvafin commented on tests/layer_tests/pytorch_tests/test_rad2deg.py:
```diff
@@ -0,0 +1,26 @@
+    def create_model(self):
+        class aten_rad2deg(torch.nn.Module):
+            def forward(self, x):
+                return x * (180.0 / np.pi)
```
```suggestion
                return torch.rad2deg(x)
```

mvafin commented on tests/layer_tests/pytorch_tests/test_rad2deg.py:
```diff
@@ -0,0 +1,26 @@
+        ref_net = None  
+        return aten_rad2deg(), ref_net, "aten::mul"
+
+    @pytest.mark.parametrize("input_shape", [(2, 2), (3, 3), (4, 5)])
```
You can test only one shape, but please add more data types for x

leoheim commented:
@mvafin 
Thank you for the feedback! I have updated the test to use only one shape (`(3, 3)`) and added support for multiple data types (`np.float32`, `np.float64`, `np.int32`).

mvafin commented on src/frontends/pytorch/src/op/rad2deg.cpp:
```diff
@@ -0,0 +1,42 @@
+    const double pi_val = std::atan(1.0) * 4;
+
+    // Create a constant node with the conversion factor (180 / π)
+    auto conversion_factor = context.mark_node(ov::op::v0::Constant::create(input_type, Shape{}, {180.0 / pi_val}));
```
Do not use input element_type to create a constant. The input type can be unknown (dynamic), create constant with fp64 type and use ConvertLIke to convert constant to required type.

leoheim commented:
Hi @mvafin ,

I have updated the code as requested.

mvafin commented:
Please fix code style:

```
src/frontends/pytorch/src/op/rad2deg.cpp:9:-#include "openvino/op/multiply.hpp"
src/frontends/pytorch/src/op/rad2deg.cpp:10:+#include "openvino/op/multiply.hpp"
src/frontends/pytorch/src/op/rad2deg.cpp:31:-    auto conversion_factor_fp64 = context.mark_node(ov::op::v0::Constant::create(ov::element::f64, Shape{}, {180.0 / pi_val}));
src/frontends/pytorch/src/op/rad2deg.cpp:31:+    auto conversion_factor_fp64 =
src/frontends/pytorch/src/op/rad2deg.cpp:32:+        context.mark_node(ov::op::v0::Constant::create(ov::element::f64, Shape{}, {180.0 / pi_val}));
src/frontends/pytorch/src/op/rad2deg.cpp:34:-    auto conversion_factor = context.mark_node(std::make_shared<ov::op::v1::ConvertLike>(conversion_factor_fp64, input));
src/frontends/pytorch/src/op/rad2deg.cpp:35:+    auto conversion_factor =
src/frontends/pytorch/src/op/rad2deg.cpp:36:+        context.mark_node(std::make_shared<ov::op::v1::ConvertLike>(conversion_factor_fp64, input));
```

mvafin commented on src/frontends/pytorch/src/op/rad2deg.cpp:
```diff
@@ -0,0 +1,48 @@
+    // Retrieve the input tensor
+    auto input = context.get_input(0);
+
+    // Get the input element type dynamically
+    auto input_type = input.get_element_type();
+
```
```suggestion
```

mvafin commented on src/frontends/pytorch/src/op/rad2deg.cpp:
```diff
@@ -0,0 +1,45 @@
+        context.mark_node(ov::op::v0::Constant::create(ov::element::f64, Shape{}, {180.0 / pi_val}));
+
+    // Convert the constant to the same type as the input tensor using ConvertLike
+    auto conversion_factor =
+        context.mark_node(std::make_shared<ov::op::v1::ConvertLike>(conversion_factor_fp64, input));
```
Test for integer case fails due to type difference. Converting to input type is not correct for integer case. Use `align_eltwise_input_types` with `input` and `conversion_factor_fp64`, leave the constant fp64, but pass true as if the second input is python scalar, that way it will scale down to fp32 both input and scalar, not fp64.

leoheim commented:
@mvafin I have updated the code as requested. The constant remains as `fp64`, and I used `align_eltwise_input_types` with the input and the constant, passing `true` to handle integer cases correctly. Let me know if there's anything else to adjust. Thanks!

mvafin commented on src/frontends/pytorch/src/op/rad2deg.cpp:
```diff
@@ -0,0 +1,44 @@
+        context.mark_node(ov::op::v0::Constant::create(ov::element::f64, Shape{}, {180.0 / pi_val}));
+
+    // Align elementwise input types to handle integer cases correctly
+    auto aligned_inputs = align_eltwise_input_types(input, conversion_factor_fp64, true);
```
This is no right way to use this function. Your code cannot be built

mvafin commented on src/frontends/pytorch/src/op/rad2deg.cpp:
```diff
@@ -0,0 +1,44 @@
+        context.mark_node(ov::op::v0::Constant::create(ov::element::f64, Shape{}, {180.0 / pi_val}));
+
+    // Align elementwise input types to handle integer cases correctly
+    align_eltwise_input_types(context, input, conversion_factor_fp64, true, false);
```
```suggestion
    align_eltwise_input_types(context, input, conversion_factor_fp64, is_python_scalar_input(context, 0), true);
```
Also fix code style, please

leoheim commented:
@mvafin 

Added static_cast for type consistency.

mvafin commented on src/frontends/pytorch/src/op/rad2deg.cpp:
```diff
@@ -0,0 +1,47 @@
+    const double pi_val = std::atan(1.0) * 4;
+
+    // Create a constant node with the conversion factor (180 / π) using fp32 type
+    auto conversion_factor_fp32 = context.mark_node(
+        ov::op::v0::Constant::create(ov::element::f32, Shape{}, {180.0f / static_cast<float>(pi_val)}));
+
+    // Convert the constant to ov::Output<ov::Node>
+    auto conversion_factor_output = ov::Output<ov::Node>(conversion_factor_fp32);
```
```suggestion
    Output<Node> conversion_factor_output = context.mark_node(
        ov::op::v0::Constant::create(ov::element::f32, Shape{}, {180.0f / static_cast<float>(pi_val)}));
```

mvafin commented:
build_jenkins

leoheim commented:
Hi @mvafin 

Some checks failed, but I don’t have access to the details. Could you help me proceed?
